File: DSA/Backtracking/Code/Subsets.cs
// Subsets using Backtracking in C#

using System;
using System.Collections.Generic;

class Subsets {
    static void PrintSubset(int[] arr, int idx, List<int> subset) {
        if (idx == arr.Length) {
            Console.Write("[");
            for (int i = 0; i < subset.Count; i++)
                Console.Write(subset[i] + (i+1==subset.Count?"":", "));
            Console.WriteLine("]");
            return;
        }
        // Include arr[idx]
        subset.Add(arr[idx]);
        PrintSubset(arr, idx+1, subset);
        subset.RemoveAt(subset.Count-1);
        // Exclude arr[idx]
        PrintSubset(arr, idx+1, subset);
    }
    static void Main() {
        Console.WriteLine("=== Subsets (C#) ===\n");
        int[] arr = {1, 2, 3};
        PrintSubset(arr, 0, new List<int>());
        Console.WriteLine("\nTime Complexity: O(2^n * n)");
        Console.WriteLine("Space Complexity: O(n)");
    }
}

File: DSA/Backtracking/Code/Subsets.c
#include <stdio.h>

// Subsets using Backtracking
void printSubset(int arr[], int n, int idx, int subset[], int subsetSize) {
    if (idx == n) {
        printf("[");
        for (int i = 0; i < subsetSize; i++) printf("%d%s", subset[i], i+1==subsetSize?"":", ");
        printf("]\n");
        return;
    }
    // Include arr[idx]
    subset[subsetSize] = arr[idx];
    printSubset(arr, n, idx+1, subset, subsetSize+1);
    // Exclude arr[idx]
    printSubset(arr, n, idx+1, subset, subsetSize);
}

int main() {
    printf("=== Subsets (C) ===\n\n");
    int arr[] = {1, 2, 3};
    int n = 3;
    int subset[10];
    printSubset(arr, n, 0, subset, 0);
    printf("\nTime Complexity: O(2^n * n)\nSpace Complexity: O(n)\n");
    return 0;
}
